### PR TITLE
[JAX] Add BRCM support for THD

### PR DESCRIPTION
**Pull request by KshitijLakhani**
# Description

Add support for Padding Bottom Right Causal Mask support for THD data types

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

1. Add support for correctly calculating the seqlens and offsets for THD Padding BRCM. 
TODO: Padding BRCM support for the fast path needs to be evaluated and added (`_segment_ids_pos_to_seqlens_offsets_fast_causal_path`).

2. Test cases added for PBRCM
3. Modified one of the test cases from `2-2048-1024-12-12-64-64-BF16-CROSS` to `2-512-1024-12-12-64-64-BF16-CROSS` for two reasons : i) To test smaller matrix size thereby reducing CI test time and ii) To test a sequence where q_len < k_len

**Additional Test timing:** 
Approximately 4.5 mins of extra test time is expected to be added (on H100) due to the newly added PBRCM test cases (~100 over different combinations)

**Profile numbers (SWA and non SWA):**
Comparing Causal mask (CM) vs Padding Bottom Right Causal mask (PBRCM) perf metrics for the JIT-ted fused DPA call made in the test (over 20 iterations).
SWA and No SWA reported separately to account for the fast path calculation difference between SWA and non SWA for CM case. 

| Mask | Random middle iter (us) |
| :------- |  -------: |
| PBRCM No SWA  | ~170  |
| CM No SWA | ~150  |
| PBRCM SWA  | ~170  |
| CM SWA  | ~150  |

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/jax/attention.py [hunk 1]
```diff
@@ -420,6 +438,33 @@ def _segment_ids_pos_to_seqlens_offsets_fast_causal_path(
...
+    """
+    boundary = jnp.concatenate(
+        [jnp.broadcast_to(True, (1,)), segment_ids_1d[1:] != segment_ids_1d[:-1]]
+    )
+    run_ids = jnp.cumsum(boundary) - 1
+    max_runs = segment_ids_1d.shape[-1]  # each element could, in worst case, start a run
+    counts = jnp.bincount(run_ids, length=max_runs)
+    # Fill in the missing values
+    segment_ids_run_length_1d = counts[run_ids]
+    segment_ids_run_length_1d = jnp.where(segment_ids_1d == 0, 0, segment_ids_run_length_1d)
```
Comment (huanghua1994) [inline transformer_engine/jax/attention.py]: Remind me, `segment_ids_1d == 0` means this location is a padding, right?

Comment (KshitijLakhani) [reply transformer_engine/jax/attention.py]: That is right.
Actually, thanks for bringing this up and adding your general comment for an example - I'll just add this simple example in the doc string for ease of understanding this code in the future
